src/v/kafka/server/handlers/delete_records.cc

```diff
@@ -0,0 +1,248 @@
+
+static result_t make_partition_error(const model::ntp& ntp, error_code err) {
+    return std::make_tuple(
+      ntp.tp.topic,
+      delete_records_partition_result{
+        .partition_index = ntp.tp.partition,
+        .low_watermark = invalid_low_watermark,
+        .error_code = err});
+}
+
+static result_t
+make_partition_response(const model::ntp& ntp, model::offset low_watermark) {
+    return std::make_tuple(
+      ntp.tp.topic,
+      delete_records_partition_result{
+        .partition_index = ntp.tp.partition,
+        .low_watermark = low_watermark,
+        .error_code = error_code::none});
+}
+
+/// If validation passes, attempts to prefix truncate the raft log at the given
+/// offset. Returns a response that includes the new low watermark
+static ss::future<result_t> prefix_truncate(
+  cluster::partition_manager& pm,
+  model::ntp ntp,
+  model::offset truncation_offset,
+  std::chrono::milliseconds timeout_ms) {
+    auto partition = pm.get(ntp);
+    if (!partition) {
+        co_return make_partition_error(
+          ntp, error_code::unknown_topic_or_partition);
+    }
+    if (!partition->is_leader()) {
+        co_return make_partition_error(
+          ntp, error_code::not_leader_for_partition);
+    }
+    if (truncation_offset == current_high_watermark) {
+        /// User is requesting to truncate all data
+        truncation_offset = partition->high_watermark();
+    }
+    if (truncation_offset < model::offset(0)) {
+        co_return make_partition_error(ntp, error_code::offset_out_of_range);
+    }
+
+    /// Perform truncation at the requested offset. A special batch will be
+    /// written to the log, eventually consumed by replicas via the
+    /// new_log_eviction_stm, which will perform a prefix truncation at the
+    /// given offset
+    auto errc = co_await partition->prefix_truncate(
+      truncation_offset, ss::lowres_clock::now() + timeout_ms);
+    if (errc) {
+        vassert(
+          errc.category() == cluster::error_category(),
+          "Unknown error category observed");
+        auto kerr = error_code::unknown_server_error;
+        switch (cluster::errc(errc.value())) {
+        case cluster::errc::timeout:
+        case cluster::errc::shutting_down:
+            kerr = error_code::request_timed_out;
+            break;
+        case cluster::errc::invalid_truncation_offset:
+            kerr = error_code::offset_out_of_range;
+            break;
+        case cluster::errc::topic_invalid_config:
+        default:
+            kerr = error_code::unknown_server_error;
+        }
+        vlog(
+          klog.info,
+          "Possible failed attempted to truncate partition: {} error: {}",
+          ntp,
+          kerr);
+        co_return make_partition_error(ntp, kerr);
+    }
+    const auto kafka_start_offset
+      = partition->get_offset_translator_state()->from_log_offset(
+        partition->start_offset());
```

Comment:
Using replicated partition would provide consistent interface it would be enough to call `start_offset` method.